rafthttp/stream.go

```diff
@@ -0,0 +1,279 @@
+package rafthttp
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+)
+
+var (
+	streamBufSize = 4096
+	errHeartbeat  = errors.New("the message is heartbeat")
+)
+
+type encoder interface {
+	// isEllegible returns whether the message is ellegible to be encoded.
+	// You must ensure the message is ellegible before calling encodeTo.
+	isEllegible(m raftpb.Message) bool
```

Comment:
It is used to check whether the message can be encoded by the encoder. It is called when deciding which one should take charge of sending this message, which happens before encodeTo. This is a non-blocking function, so it will not block the decision process.